src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java

```diff
@@ -63,17 +80,36 @@
 			throw new IllegalArgumentException(message);
 		}
 
-		Annotation argumentsSource = argumentsSources.get(0);
-		Class<? extends Annotation> argumentsSourceClass = argumentsSource.annotationType();
-		Class<? extends Range> rangeClass = argumentsSourceClass.getAnnotation(RangeClass.class).value();
+		argumentsSource = argumentsSources.get(0);
+	}
 
-		Range<?> range = (Range) rangeClass.getConstructors()[0].newInstance(argumentsSource);
-		range.validate();
-		return asStream(range).map(Arguments::of);
+	private void verifyNoContainerAnnotationIsPresent(AnnotatedElement element) {
+		if (Stream.of(element.getAnnotations()).anyMatch(this::isContainerAnnotation))
+			throw new IllegalArgumentException(
+				"Range source annotation should not be repeated for @ParameterizedTest. @ParameterizedTest should have exactly one argument source.");
+	}
+
+	private boolean isContainerAnnotation(Annotation annotation) {
+		// See https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.6.3
+		return Stream
+				.of(annotation.annotationType().getMethods())
+				.anyMatch(method -> method.getName().equals("value") && method.getReturnType().isArray()
+						&& method.getReturnType().getComponentType().isAnnotation()
+						&& declaresContainer(method.getReturnType().getComponentType(), annotation));
 	}
 
-	private Stream<?> asStream(Range r) {
+	private static boolean declaresContainer(Class<?> componentType, Annotation annotation) {
+		Repeatable repeatable = componentType.getAnnotation(Repeatable.class);
+		return repeatable != null && repeatable.value().equals(annotation.annotationType());
+	}
```

Comment:
I just saw this code in `PioneerAnnotationUtils`, no need to repeat it here. Just expose there and add (reflectively) to the `params` variant.